### PR TITLE
Query module improvements

### PR DESCRIPTION
**Pull request by imilinovic**
Performance enhancements:
- C++ API `MemoryDispatcherGuard` had an expensive registration and fetch mechanism, replaced with thread_local state
- Argument construction made temporary buffers unnecessarily
- Hot-loading required module lookup on every call, we modified implementation to keep module lifetime for the whole function usage

### CODE REVIEW

File: include/mgp.hpp
Comment (antoniofilipovic) [inline include/mgp.hpp:132]: What if a user creates multiple threads in QM? In that case, user would need to GetMemoryResource from main thread and register MemoryResource on each thread created in QM, right? But at that point, we have I guess different concurrency issue on that one mgp_memory? I think that issue is expected as in memory.cpp we don't have any guarantees for multiple threads accessing the same memory resource. So we should just advise users not to do such things in QMs?

**Comment by Ignition** [comment]
### Tracking
- [ ] **[Link to Epic/Issue]**


### Standard development
- [ ] Update unit/E2E tests
- [ ] Compare the [benchmarking results](https://bench-graph.memgraph.com/) between the master branch this branch


### CI Testing Labels
- [ ] Select the appropriate CI test labels _(CI -build=**build-name** -test=**test-suite**)_


### Documentation checklist
- [ ] Add the documentation label
- [ ] Add the bug / feature label
- [ ] Add the milestone for which this feature is intended
    - If not known, set for a later milestone
- [ ] Write a release note, including added/changed clauses
    - **[ Release note text ]**
- [ ] **[ Documentation PR link memgraph/documentation#XXXX ]**
    - [ ] Is back linked to this development PR
- [ ] **[ Tag someone from docs team ]**

File: include/mgp.hpp
Comment (Ignition) [reply include/mgp.hpp]: ATM any usage of mgp_memory is not thread safe. We will handle that when a user requires concurrent usage of mgp_memory

Comment (antoniofilipovic) [reply include/mgp.hpp]: @Ignition I would then update the docs to let users know this will not work, maybe here: https://memgraph.com/docs/custom-query-modules/manage-query-modules

File: src/query/plan/operator.cpp
Comment (antoniofilipovic) [inline src/query/plan/operator.cpp:5202]: I think you don't need this check, since the module will be nullptr always on the next pull
```suggestion
      
      auto maybe_found = procedure::FindProcedure(procedure::gModuleRegistry, self_->procedure_name_,
                                                  context.evaluation_context.memory);
      if (!maybe_found) {
        throw QueryRuntimeException("There is no procedure named '{}'.", self_->procedure_name_);
      }

      module = std::move(maybe_found->first);
      proc = maybe_found->second;

      if (proc->info.is_write != self_->is_write_) {
        auto get_proc_type_str = [](bool is_write) { return is_write ? "write" : "read"; };
        throw QueryRuntimeException("The procedure named '{}' was a {} procedure, but changed to be a {} procedure.",
                                    self_->procedure_name_, get_proc_type_str(self_->is_write_),
                                    get_proc_type_str(proc->info.is_write));
      }
      

```

**Review by antoniofilipovic — CHANGES_REQUESTED** [review]
Few comments, but very nice

File: src/query/procedure/callable_alias_mapper.hpp
Comment (Ignition) [reply src/query/procedure/callable_alias_mapper.hpp:38]: To avoid needing to allocate `std::string` for lookups. I prefer to pass around `std::string_view`. We could use a heterogeneous hash function for strings if needed, but for now I am happy to move from unordered_map to map, until a performance related issue highlights that it needs to be a unordered_map.

File: src/query/plan/operator.cpp
Comment (Ignition) [reply src/query/plan/operator.cpp:5186]: This no longer applies. The ModulePtr in the past use to lock the registry (this IMO was too strong), hence this comment to avoid deadlock. 

We now prevent replacing query module while it is in use. The shared_ptr's use_count() != 1 means at least one query is using it.

File: src/query/plan/operator.cpp
Comment (Ignition) [reply src/query/plan/operator.cpp:5202]: It is possible to go around this loop multiple times.
- First call to procedure could give 0 rows.
- hence we go around again, no need to FindProcedure again

File: src/query/plan/operator.cpp
Comment (antoniofilipovic) [reply src/query/plan/operator.cpp:5186]: Aha yes, removing is prevented due to check

File: src/query/plan/operator.cpp
Comment (antoniofilipovic) [reply src/query/plan/operator.cpp]: You are right, in that case it will try to pull again and do the processing, makes sense

**Comment by antoniofilipovic** [comment]
@Ignition re-request review when it is ready to be reviewed again, I resolved older comments

File: src/query/plan/operator.cpp
Comment (antoniofilipovic) [inline src/query/plan/operator.cpp:5152]: This would be a breaking change in my opinion. If we cache, this means the user can't reload the module. It also means that such queries that reload a module will block the the memgraph's other queries that want to take a shared lock and execute queries with procedures. I am against this change honestly

**Review by antoniofilipovic — APPROVED** [review]
Good work!

**Comment by MarkoBarisic** [comment]
Don't merge before I approve, there have been changes in `mgbuild.sh`, `jepsen/run.sh` and `diff_jepsen.yaml`

**Review by MarkoBarisic — APPROVED** [review]
LGTM

**Pull request merged**
